### DISCUSSION TITLE
Initial value problem

### DISCUSSION THREAD

**Discussion by maeckha** [CliMA/Oceananigans.jl#1922] [General]
1. Set up an initial value problem without particles, including visualization. Inspect the flow, tweak parameters, and develop an intuition for how changes to parameters change the resulting flow. Settle on a setup that you think will produce something sensible.
2. Add a very simple implementation of Lagrangian particles that are passively advected, with no “dynamics” / rock papers scissor rule. Set up visualization that combines the flow visuzliation with particle trajectors. Iterate on the setup so that the visualization is helpful and informative.
3. Implement incremental changes to the flow or particle dynamics that get you closer to the problem you’d like to simulate ultimately.

Per the approach above I have started with a uniform condition.

```julia
grid = RegularRectilinearGrid(size = (16, 16, 1), extent = (2π, 2π, 1))
model = NonhydrostaticModel(grid = grid)
set!(model, u=0.1, v=1.5)
```

What would the initial condition have to be, if a laminar flow was to enter left and leave right?

**Comment by glwagner** [comment]
Try

```julia
grid = RegularRectilinearGrid(size = (128, 128), x=(-5, 5), y=(-5, 5), topology=(Periodic, Periodic, Flat))
model = NonhydrostaticModel(grid = grid, tracers=:c, buoyancy=nothing)

initial_c(x, y, z) = exp(-x^2 - y^2)
set!(model, u=1, c=initial_c)
```

this initializes a uniform velocity from left to right, with a Gaussian tracer profile:

```julia
using UnicodePlots
heatmap(interior(model.tracers.c)[:, :, 1]')
```

producing

![image](https://user-images.githubusercontent.com/15271942/128071996-23b42bbd-51ca-4082-96ce-fba7026fc12c.png)

stepping the model forward shows the effect of advection by a uniform flow on the tracer distribution:

```julia
simulation = Simulation(model, Δt=1e-2, stop_iteration=100)
run!(simulation)
heatmap(interior(model.tracers.c)[:, :, 1]')
```

![image](https://user-images.githubusercontent.com/15271942/128072220-56e1003a-a1ac-4414-876e-e56dffd8984b.png)

**Reply by maeckha** [reply]
Thank you for the example. 
Will it be possible to add particles here and visualize them with heatmap as well?

**Reply by glwagner** [reply]
I think so but I'm not very familiar with the particles interface. @ali-ramadhan would be able to answer that better. I don't think you visualize particle trajectories with `heatmap`. `heatmap` makes sense when the thing you're visualizing has a value at every grid point (like tracer concentration). A particle position is just a single coordinate. I think you just want to use `plot` for that.

**Reply by maeckha** [reply]
I am sorry, if this is a rather naive question, but how should I change the function or the forward stepping so that the effect of advection starts at the left edge and travels all the way to the right?

**Reply by glwagner** [reply]
Are you saying that you want `u=-1` rather than `u=1`?

**Reply by maeckha** [reply]
Sorry for not putting this more clearly. My question was not about u = -1. I meant how do I get the function to move from left to right in multiple time steps.
I think I would have to pass a timestepper like runge-kutta 3 to the model constructor?

@ali-ramadhan , creating the particles like in the docs seems pretty straight forward like this:

```
n_particles = 10;

x₀ = zeros(n_particles);
                              
y₀ = rand(n_particles);
                              
z₀ = -0.5 * ones(n_particles);
                              
lagrangian_particles = LagrangianParticles(x=x₀, y=y₀, z=z₀)
```

What I do not understand yet, is how I would plot the particles in this example. It seems quite different than your rock-paper-scissors example.

**Reply by glwagner** [reply]
This part of the code:

```julia
simulation = Simulation(model, Δt=1e-2, stop_iteration=100)
run!(simulation)
```

causes the model to step forward 100 time-steps using the `QuasiAdamsBashforth2TimeStepper`. As a result of this time-stepping, the tracer distribution moves from left to right. Try changing these two lines to

```julia
simulation = Simulation(model, Δt=1e-2, stop_iteration=400)
run!(simulation)
```

(increasing `stop_iteration` from 100 to 400) and see what happens.

**Reply by glwagner** [reply]
I believe after the particles are created, you have to add them to the model by writing

```julia
model = NonhydrostaticModel(grid = grid, tracers=:c, buoyancy=nothing, lagrangian_particles=lagrangian_particles)
```

I would also use

```julia
z₀ = zeros(n_particles)
```

**Reply by maeckha** [reply]
I have set the timesteps too small and therefore could not see enough progress in the heatmap. Stepping from 100 to 400 looks better. Don't know why I have stuck to incremental stepping for that long, thanks.

Does the 'u' here mean velocity? 
`set!(model, u=1, c=initial_c)`

**Reply by glwagner** [reply]
On a rectilinear grid, `u, v, w` are the velocity components in the `x, y, z` direction.

**Reply by maeckha** [reply]
> I believe after the particles are created, you have to add them to the model by writing
> 
> ```julia
> model = NonhydrostaticModel(grid = grid, tracers=:c, buoyancy=nothing, lagrangian_particles=lagrangian_particles)
> ```
> 
> I would also use
> 
> ```julia
> z₀ = zeros(n_particles)
> ```

```
model = NonhydrostaticModel(grid = grid, tracers=:c, buoyancy=nothing, particles=lagrangian_particles)
```
